In [179]:
import pandas as pd

In [180]:
import numpy as np

In [181]:
df=pd.read_csv('data.csv')

In [182]:
df

Suburb              Address  Rooms Type      Price Method  \
0          Abbotsford        68 Studley St    2.0    h        NaN     SS   
1          Abbotsford         85 Turner St    2.0    h  1480000.0      S   
2          Abbotsford      25 Bloomburg St    2.0    h  1035000.0      S   
3          Abbotsford   18/659 Victoria St    3.0    u        NaN     VB   
4          Abbotsford         5 Charles St    3.0    h  1465000.0     SP   
5          Abbotsford     40 Federation La    3.0    h   850000.0     PI   
6          Abbotsford          55a Park St    4.0    h  1600000.0     VB   
7          Abbotsford         16 Maugie St    4.0    h        NaN     SN   
8          Abbotsford         53 Turner St    2.0    h        NaN      S   
9          Abbotsford         99 Turner St    2.0    h        NaN      S   
10         Abbotsford       129 Charles St    2.0    h   941000.0      S   
11         Abbotsford         124 Yarra St    3.0    h  1876000.0      S   
12         Abbotsford  121/56 Nicholson St    2.0    u        NaN     PI   
13         Abbotsford        17 Raphael St    4.0    h        NaN      W   
14         Abbotsford        98 Charles St    2.0    h  1636000.0      S   
15         Abbotsford     217 Langridge St    3.0    h  1000000.0      S   
16         Abbotsford      18a Mollison St    2.0    t   745000.0      S   
17         Abbotsford   6/241 Nicholson St    1.0    u   300000.0      S   
18         Abbotsford        10 Valiant St    2.0    h  1097000.0      S   
19         Abbotsford  403/609 Victoria St    2.0    u   542000.0      S   
20         Abbotsford            2 Rich St    2.0    h        NaN     SP   
21         Abbotsford    25/84 Trenerry Cr    2.0    u   760000.0     SP   
22         Abbotsford    106/119 Turner St    1.0    u   481000.0     SP   
23         Abbotsford   411/8 Grosvenor St    2.0    u   700000.0     VB   
24         Abbotsford      40 Nicholson St    3.0    h  1350000.0     VB   
25         Abbotsford  123/56 Nicholson St    2.0    u   750000.0      S   
26         Abbotsford           22 Park St    4.0    h  1985000.0      S   
27         Abbotsford    13/84 Trenerry Cr    1.0    u   500000.0      S   
28         Abbotsford        45 William St    2.0    h  1172500.0      S   
29         Abbotsford   7/20 Abbotsford St    1.0    u   441000.0     SP   
...               ...                  ...    ...  ...        ...    ...   
23648   Vermont South        2 Settlers Ct    4.0    h  1295000.0     SP   
23649   Vermont South        9 Winswood Cl    4.0    h  1250000.0     PI   
23650        Viewbank        149 Graham Rd    5.0    h  1316000.0     SP   
23651        Viewbank       7 Northwood Dr    4.0    h        NaN     SP   
23652        Wantirna   16 chesterfield Ct    4.0    h   951000.0      S   
23653        Wantirna            9 eton Sq    4.0    h        NaN     PI   
23654  Wantirna South         12 Armagh Cr    4.0    h  1323000.0      S   
23655  Wantirna South        34 Fewster Dr    3.0    h   970000.0      S   
23656  Wantirna South           15 Mara Cl    4.0    h  1330000.0      S   
23657        Watsonia       2/14 Daours Ct    2.0    u   590000.0      S   
23658        Watsonia        76 Kenmare St    2.0    h   650000.0     PI   
23659  Watsonia North         22 Meakin St    3.0    h        NaN     PI   
23660        Werribee          6 Feijoa Ct    5.0    h        NaN      W   
23661        Werribee       11 Gundaroo Sq    4.0    h        NaN     PI   
23662        Werribee         91 Latham St    4.0    h   540000.0     PI   
23663        Werribee          5 Nuragi Ct    4.0    h   635000.0      S   
23664        Werribee             3 Sky Ct    3.0    h        NaN     PI   
23665  West Footscray      1/21 Waratah St    2.0    t   655000.0      S   
23666     Westmeadows           9 Black St    3.0    h   582000.0      S   
23667     Westmeadows      81 Hillcrest Dr    4.0    h   720000.0     VB   
23668   Wheelers Hill         12 Strada Cr    4.0    h  1245000

## Ans:1   The no. of rows and no. of columns in the dataset

In [183]:
df.shape  # Ans:1

(23678, 21)

The number of rows in data are 23678 and columns are 21.

## Ans:2   The datatypes of each column

In [184]:
df.dtypes # Ans:2

Suburb            object
Address           object
Rooms            float64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

## Ans:3   The row indexes of rows having all null values 

In [185]:
df_empty_row= df[df.isnull().all(axis=1)].index.tolist() # 
df_empty_row

[31,
 1145,
 1348,
 2119,
 3277,
 3288,
 3745,
 3752,
 3798,
 3855,
 4089,
 4096,
 4164,
 4218,
 4222,
 4326,
 4649,
 4897,
 4942,
 5052,
 5223,
 5265,
 5328,
 5331,
 5383,
 5405,
 5625,
 5741,
 5952,
 6125,
 6133,
 6140,
 6143,
 6251,
 6326,
 6406,
 6446,
 6703,
 6717,
 6853,
 6921,
 7086,
 7261,
 7348,
 7596,
 7743,
 7759,
 7762,
 7775,
 8054,
 8095,
 8118,
 8190,
 8409,
 8501,
 8652,
 8816,
 8974,
 9098,
 9210,
 9232,
 9450,
 9455,
 9586,
 9761,
 9888,
 9894,
 10009,
 10125,
 10390,
 10475,
 10738,
 10952,
 11334,
 11348,
 11552,
 11755,
 12087,
 12441,
 12559,
 12997,
 13066,
 13381,
 13647,
 13725,
 14009,
 14471,
 14644,
 14963,
 15094,
 15214,
 15771,
 16044,
 16252,
 16563,
 16655,
 16848,
 17111,
 17246,
 17611,
 17741,
 18077,
 18148,
 18251,
 18374,
 18476,
 18547,
 18671,
 18742,
 18884,
 19001,
 19287,
 19641,
 19949,
 20087,
 20422,
 20496,
 20695,
 20838,
 20880,
 21055,
 21220,
 21343,
 21602,
 22003,
 22109,
 22288,
 22704,
 22918,
 23153,
 23341]

## Ans:4    The total number of empty rows and their percentage in whole dataset and remove them

In [186]:
df_t_empty_rows= len(df[df.isnull().all(axis=1)]) # Tatal number of empty rows
df_t_empty_rows

131

In [187]:
len(df)  # Total number of rows in dataset

23678

In [188]:
pct=df_t_empty_rows/len(df) # Percentage of empty rows in the dataset

In [189]:
pct*100

0.5532561871779711

In [190]:
df1=df.dropna(how='all',axis=0) # Removing the empty rows

In [191]:
df1

Suburb              Address  Rooms Type      Price Method  \
0          Abbotsford        68 Studley St    2.0    h        NaN     SS   
1          Abbotsford         85 Turner St    2.0    h  1480000.0      S   
2          Abbotsford      25 Bloomburg St    2.0    h  1035000.0      S   
3          Abbotsford   18/659 Victoria St    3.0    u        NaN     VB   
4          Abbotsford         5 Charles St    3.0    h  1465000.0     SP   
5          Abbotsford     40 Federation La    3.0    h   850000.0     PI   
6          Abbotsford          55a Park St    4.0    h  1600000.0     VB   
7          Abbotsford         16 Maugie St    4.0    h        NaN     SN   
8          Abbotsford         53 Turner St    2.0    h        NaN      S   
9          Abbotsford         99 Turner St    2.0    h        NaN      S   
10         Abbotsford       129 Charles St    2.0    h   941000.0      S   
11         Abbotsford         124 Yarra St    3.0    h  1876000.0      S   
12         Abbotsford  121/56 Nicholson St    2.0    u        NaN     PI   
13         Abbotsford        17 Raphael St    4.0    h        NaN      W   
14         Abbotsford        98 Charles St    2.0    h  1636000.0      S   
15         Abbotsford     217 Langridge St    3.0    h  1000000.0      S   
16         Abbotsford      18a Mollison St    2.0    t   745000.0      S   
17         Abbotsford   6/241 Nicholson St    1.0    u   300000.0      S   
18         Abbotsford        10 Valiant St    2.0    h  1097000.0      S   
19         Abbotsford  403/609 Victoria St    2.0    u   542000.0      S   
20         Abbotsford            2 Rich St    2.0    h        NaN     SP   
21         Abbotsford    25/84 Trenerry Cr    2.0    u   760000.0     SP   
22         Abbotsford    106/119 Turner St    1.0    u   481000.0     SP   
23         Abbotsford   411/8 Grosvenor St    2.0    u   700000.0     VB   
24         Abbotsford      40 Nicholson St    3.0    h  1350000.0     VB   
25         Abbotsford  123/56 Nicholson St    2.0    u   750000.0      S   
26         Abbotsford           22 Park St    4.0    h  1985000.0      S   
27         Abbotsford    13/84 Trenerry Cr    1.0    u   500000.0      S   
28         Abbotsford        45 William St    2.0    h  1172500.0      S   
29         Abbotsford   7/20 Abbotsford St    1.0    u   441000.0     SP   
...               ...                  ...    ...  ...        ...    ...   
23648   Vermont South        2 Settlers Ct    4.0    h  1295000.0     SP   
23649   Vermont South        9 Winswood Cl    4.0    h  1250000.0     PI   
23650        Viewbank        149 Graham Rd    5.0    h  1316000.0     SP   
23651        Viewbank       7 Northwood Dr    4.0    h        NaN     SP   
23652        Wantirna   16 chesterfield Ct    4.0    h   951000.0      S   
23653        Wantirna            9 eton Sq    4.0    h        NaN     PI   
23654  Wantirna South         12 Armagh Cr    4.0    h  1323000.0      S   
23655  Wantirna South        34 Fewster Dr    3.0    h   970000.0      S   
23656  Wantirna South           15 Mara Cl    4.0    h  1330000.0      S   
23657        Watsonia       2/14 Daours Ct    2.0    u   590000.0      S   
23658        Watsonia        76 Kenmare St    2.0    h   650000.0     PI   
23659  Watsonia North         22 Meakin St    3.0    h        NaN     PI   
23660        Werribee          6 Feijoa Ct    5.0    h        NaN      W   
23661        Werribee       11 Gundaroo Sq    4.0    h        NaN     PI   
23662        Werribee         91 Latham St    4.0    h   540000.0     PI   
23663        Werribee          5 Nuragi Ct    4.0    h   635000.0      S   
23664        Werribee             3 Sky Ct    3.0    h        NaN     PI   
23665  West Footscray      1/21 Waratah St    2.0    t   655000.0      S   
23666     Westmeadows           9 Black St    3.0    h   582000.0      S   
23667     Westmeadows      81 Hillcrest Dr    4.0    h   720000.0     VB   
23668   Wheelers Hill         12 Strada Cr    4.0    h  1245000

In [192]:
df1.shape

(23547, 21)

## Ans:5   The column names having missing values

In [193]:
df1.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

In [194]:
df1.columns[df1.isnull().any()] # Column names having missing values

Index(['Price', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

## Ans:6   The missing values and their percentage in each column

In [195]:
missing_values=df1.isnull().sum() # Missing values
missing_values

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

In [198]:
(missing_values/len(df1))*100  # Percentage of missing values in each column

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price            21.875398
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.004247
Postcode          0.004247
Bedroom2         19.030025
Bathroom         19.042766
Car              19.645815
Landsize         26.062768
BuildingArea     57.455302
YearBuilt        50.991634
CouncilArea      33.511700
Lattitude        18.278337
Longtitude       18.278337
Regionname        0.004247
Propertycount     0.004247
dtype: float64

## Ans:7    The row indexes having missing values

In [205]:
row_indexes_missing_values=df1.loc[df1.isnull().sum(axis=1)>0].index
row_indexes_missing_values

Int64Index([    0,     1,     3,     5,     7,     8,     9,    10,    12,
               13,
            ...
            23668, 23669, 23670, 23671, 23672, 23673, 23674, 23675, 23676,
            23677],
           dtype='int64', length=17351)

## Ans:8   The missing values and their percentage in each row

In [206]:
df1.isnull().sum(axis=1)  # Missing values in each row

0        3
1        2
2        0
3        3
4        0
5        2
6        0
7        1
8        2
9        2
10       2
11       0
12       1
13       1
14       0
15       9
16       9
17       2
18       0
19       9
20       1
21       9
22       9
23       2
24       0
25       0
26       7
27       9
28       2
29       2
        ..
23648    2
23649    3
23650    3
23651    2
23652    1
23653    4
23654    3
23655    3
23656    1
23657    2
23658    1
23659    5
23660    4
23661    5
23662    9
23663    1
23664    5
23665    2
23666    3
23667    4
23668    2
23669    1
23670    2
23671    2
23672    1
23673    2
23674    8
23675    4
23676    1
23677    2
Length: 23547, dtype: int64

In [207]:
df1.isnull().mean(axis=1).round(4)*100  # Percentage of missing values in each row

0        14.29
1         9.52
2         0.00
3        14.29
4         0.00
5         9.52
6         0.00
7         4.76
8         9.52
9         9.52
10        9.52
11        0.00
12        4.76
13        4.76
14        0.00
15       42.86
16       42.86
17        9.52
18        0.00
19       42.86
20        4.76
21       42.86
22       42.86
23        9.52
24        0.00
25        0.00
26       33.33
27       42.86
28        9.52
29        9.52
         ...  
23648     9.52
23649    14.29
23650    14.29
23651     9.52
23652     4.76
23653    19.05
23654    14.29
23655    14.29
23656     4.76
23657     9.52
23658     4.76
23659    23.81
23660    19.05
23661    23.81
23662    42.86
23663     4.76
23664    23.81
23665     9.52
23666    14.29
23667    19.05
23668     9.52
23669     4.76
23670     9.52
23671     9.52
23672     4.76
23673     9.52
23674    38.10
23675    19.05
23676     4.76
23677     9.52
Length: 23547, dtype: float64

## Ans:9   The column names having more than 30% missing values and remove them from the dataset

In [ ]:
# Column names having more than 30% missing values

In [210]:
col_miss=df1.isnull().sum()/len(df1)  
na_features=col_miss[col_miss > 0.30].index
na_features

Index(['BuildingArea', 'YearBuilt', 'CouncilArea'], dtype='object')

In [211]:
df1.shape

(23547, 21)

In [212]:
# Now removing columns having more 30% missing values
df1.drop(na_features, axis=1, inplace=True)

In [213]:
df1.shape

(23547, 18)

## Ans:10  The number and percentage of rows having more than 5 missing values. If the percentage of such rows is less than 20% then remove all these rows

In [217]:
row_missing=df1.loc[df1.isnull().sum(axis=1)>5].index  # Number of rows having more than 5 missing values.

In [218]:
row_missing

Int64Index([   15,    16,    19,    21,    22,    27,    34,    35,    46,
               47,
            ...
            23573, 23577, 23580, 23584, 23603, 23626, 23630, 23633, 23639,
            23662],
           dtype='int64', length=4278)

In [219]:
len(row_missing)

4278

In [221]:
(len(row_missing)/len(df1))*100

18.16791948018856

The percentage of such rows is less than 20% so we will remove all these rows

In [222]:
df1.drop(df1[df1.isnull().sum(axis=1)>5].index,inplace=True)

In [223]:
df1.shape

(19269, 18)

## Ans:11   The description of data and percentage of missing values still left column wise

In [224]:
df1.info() # Description of data

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19269 entries, 0 to 23677
Data columns (total 18 columns):
Suburb           19269 non-null object
Address          19269 non-null object
Rooms            19269 non-null float64
Type             19269 non-null object
Price            15086 non-null float64
Method           19269 non-null object
SellerG          19269 non-null object
Date             19269 non-null object
Distance         19269 non-null float64
Postcode         19269 non-null float64
Bedroom2         19066 non-null float64
Bathroom         19063 non-null float64
Car              18921 non-null float64
Landsize         17410 non-null float64
Lattitude        19243 non-null float64
Longtitude       19243 non-null float64
Regionname       19269 non-null object
Propertycount    19269 non-null float64
dtypes: float64(11), object(7)
memory usage: 2.8+ MB


In [228]:
(df1.isnull().sum()/len(df1))*100 # percentage of missing values still left column wise

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price            21.708444
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.000000
Postcode          0.000000
Bedroom2          1.053506
Bathroom          1.069075
Car               1.806010
Landsize          9.647621
Lattitude         0.134932
Longtitude        0.134932
Regionname        0.000000
Propertycount     0.000000
dtype: float64

## Ans:12    Considering the 'Price' column as target variable which is to be predicted, how will you treat the missing values? Give a brief explanation and treat the missing values in 'Price' column

In [229]:
df1['Price'].isnull().sum()

4183

In [232]:
(df1['Price'].isnull().sum()/len(df1))*100

21.70844361409518

We can not drop 21% of null values, so we will treat the missing value by fillna method using mean 

In [233]:
df1['Price'].fillna(df1['Price'].mean(),inplace=True)

C:\Users\tiya\Anaconda3\lib\site-packages\pandas\core\generic.py:6130: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


In [234]:
df1['Price'].isnull().sum()

0

Now we do not have any null values in the Price column.

## Ans:13   Now check the percentage of missing values in each column again, if missing values exist, treat it with the most suitable method

In [235]:
(df1.isnull().sum()/len(df1))*100

Suburb           0.000000
Address          0.000000
Rooms            0.000000
Type             0.000000
Price            0.000000
Method           0.000000
SellerG          0.000000
Date             0.000000
Distance         0.000000
Postcode         0.000000
Bedroom2         1.053506
Bathroom         1.069075
Car              1.806010
Landsize         9.647621
Lattitude        0.134932
Longtitude       0.134932
Regionname       0.000000
Propertycount    0.000000
dtype: float64

##### Here, still missing values exist, so we will treat them.

In [236]:
df1.columns[df1.isnull().any()]

Index(['Bedroom2', 'Bathroom', 'Car', 'Landsize', 'Lattitude', 'Longtitude'], dtype='object')

In [239]:
# treating the null values
for col in ['Bedroom2', 'Bathroom', 'Car', 'Landsize', 'Lattitude', 'Longtitude']:
    df1[col].fillna(df1[col].mean(), inplace=True)

In [240]:
df1.isnull().sum()

Suburb           0
Address          0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
Lattitude        0
Longtitude       0
Regionname       0
Propertycount    0
dtype: int64

#### Now, no null values there in the dataset

## Ans:14  The number and percentage of columns removed

In [241]:
df.shape

(23678, 21)

In [242]:
df1.shape

(19269, 18)

#### We removed 3 columns from the dataset.

In [243]:
# Now it's percentage
(3/21)*100

14.285714285714285

## Ans:15   The number and percentage of rows removed in whole process

In [244]:
len(df)

23678

In [245]:
len(df1)

19269

In [246]:
len(df)-len(df1)

4409

In [248]:
# It's percentage
((len(df)-len(df1))/len(df))*100

18.620660528760872

We removed 18 % of rows from the data

## We removed 18.62% of data from the dataset, but if we remove more than 10% of data it will not appropriate. In this dataset we lost some imformation about data.